plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php

```diff
@@ -0,0 +1,186 @@
+		);
+	}
+
+	/**
+	 * Builds a SQL statement to either count or obtain IDs for orders in need of cleanup.
+	 *
+	 * @param array   $order_ids If provided, the query will only include orders in this set of order IDs or ID ranges (like "10-100").
+	 * @param string  $result    Use 'count' to build a query that returns a count. Otherwise, the query will return order IDs.
+	 * @param integer $limit     If provided, the query will be limited to this number of results. Does not apply when $result is 'count'.
+	 * @return string SQL query.
+	 */
+	private function build_sql_query_for_cleanup( array $order_ids = array(), string $result = 'ids', int $limit = 0 ): string {
+		global $wpdb;
+
+		$sql_where = '';
+
+		if ( $order_ids ) {
+			// Expand ranges in $order_ids as needed to build the WHERE clause.
+			$where_ids    = array();
+			$where_ranges = array();
+
+			foreach ( $order_ids as &$arg ) {
+				if ( is_numeric( $arg ) ) {
+					$where_ids[] = absint( $arg );
+				} elseif ( preg_match( '/^(\d+)-(\d+)$/', $arg, $matches ) ) {
+					$where_ranges[] = $wpdb->prepare( "({$wpdb->posts}.ID >= %d AND {$wpdb->posts}.ID <= %d)", absint( $matches[1] ), absint( $matches[2] ) );
+				}
+			}
+
+			if ( $where_ids ) {
+				$where_ranges[] = "{$wpdb->posts}.ID IN (" . implode( ',', $where_ids ) . ')';
+			}
+
+			if ( ! $where_ranges ) {
+				$sql_where .= '1=0';
+			} else {
+				$sql_where .= '(' . implode( ' OR ', $where_ranges ) . ')';
+			}
+		}
+
+		$sql_where .= $sql_where ? ' AND ' : '';
+		$sql_where .= "{$wpdb->posts}.post_type IN ('" . implode( "', '", esc_sql( wc_get_order_types( 'cot-migration' ) ) ) . "')";
+
+		if ( 'count' === $result ) {
+			$sql_fields = 'COUNT(*)';
+			$sql_limit  = '';
+		} else {
+			$sql_fields = 'ID';
+			$sql_limit  = $limit > 0 ? $wpdb->prepare( 'LIMIT %d', $limit ) : '';
+		}
+
+		return "SELECT {$sql_fields} FROM {$wpdb->posts} WHERE {$sql_where} {$sql_limit}";
+	}
+
+	/**
+	 * Performs a cleanup of post data for a given order and also converts the post to the placeholder type in the backup table.
+	 *
+	 * @param int $order_id Order ID.
+	 * @return void
+	 * @throws \Exception When an error occurs.
+	 */
+	public function cleanup_post_data( $order_id ): void {
+		global $wpdb;
+
+		$order = wc_get_order( $order_id );
+		if ( ! $order ) {
+			// translators: %d is an order ID.
+			throw new \Exception( sprintf( __( '%d is not a valid order ID.', 'woocommerce' ), $order_id ) );
+		}
+
+		if ( ! $this->is_order_newer_than_post( $order ) ) {
+			throw new \Exception( sprintf( __( 'Data in posts table appears to be more recent than in HPOS tables.', 'woocommerce' ) ) );
+		}
+
+		$meta_ids = $wpdb->get_col( $wpdb->prepare( "SELECT meta_id FROM {$wpdb->postmeta} WHERE post_id = %d", $order->get_id() ) );
+		foreach ( $meta_ids as $meta_id ) {
+			delete_metadata_by_mid( 'post', $meta_id );
+		}
+
+		$wpdb->query(
+			$wpdb->prepare(
+				"UPDATE {$wpdb->posts} SET post_type = %s, post_status = %s WHERE ID = %d",
```

Comment:
Mostly just curious, but a few lines above we use the core WP function `delete_metadata_by_mid()` which allows other code to observe the delete event (because of the related core hooks that will be fired). Here, we use a direct query and the modifications are essentially unobservable. Is that deliberate, vs using `wp_update_post()`?

Reply:
I was just trying to isolate the operation a bit. My reasoning was that `wp_update_post()` fires a bunch of hooks (including `save_post`) so was more likely to result in hypothetical 3rd party code "reacting" to the event (by doing something, which could include adding metadata back or something). I guess this could be considered unlikely, so happy to use `wp_update_post()` if you think that would be better.

Reply:
Both seem reasonable. I'm a little unsure what would be best: it just seemed asymmetric in the sense that we use core WP functionality for the meta data itself (which does trigger various hooks) vs how we update the post. I think you are on to something, though. 

If we stick with the current approach, maybe we should record why we made this decision with a comment *('we suspect using wp_update_post could lead to integrations mistakenly updating the entity'* or something along those lines)?

Reply:
Fair enough. I'll add a comment. I guess another benefit to _not_ using `wp_update_post_meta()` is we can skip the change to "post modified" that it'd trigger, which isn't correct in this case.